scalableminds/webknossos
# Colors for mappings

This PR adds the possibility to add custom colors for mappings as well as to hide unmapped segments.
Up to 256 colors (hsv hue values) can be specified which are then used in order to color the equivalence classes of the mapping. The interface using the API is not as clean yet (see my description in https://discuss.webknossos.org/t/colors-for-mappings/602/27?u=dwerner) but as the single user of that feature will be using the json mapping interface, I think that is fine for now. We'll hopefully align the json and API mappings interface in the near future.

Note that the are now two ways that segments can end up to be transparent. If the user specifies the `hideUnmappedIds` flag and doesn't map a segment id, it will be transparent. Also if the user specifies not enough entries in the `mappingColors` array (e.g.: 3 equivalence classes, but only 2 hue values), the segments of the equivalence classes without a color will be transparent.
Thinking about it while writing this PR description, maybe it would be better to have a default color in that second case. If I remember correctly I introduced this behavior in the beginning to solve the `hideUnmappedIds` feature, which I solved differently now ^^

### URL of deployed dev instance (used for testing):
- https://colorsformappings.webknossos.xyz

### Steps to test:
- Open a tracing using the ROI2017_wkw dataset.
  - Use this user script to activate a mapping with custom colors:
  ```javascript
  webknossos.apiReady(3).then(api => {
  	const segmentationLayerName = api.data.getVolumeTracingLayerName();
  	api.data.setMapping(
  		segmentationLayerName,
  		{ 1230341: 1, 1230557: 1, 1227259: 2, 1236582: 2 },
      	        { colors: [0.0, 0.2, 0.5], hideUnmappedIds: true }
  	);
  });
  ```
  - All unmapped ids should be transparent, and the mapped ids should have the specified colors.

- Locally, add a mapping file like this in the mappings folder of the segmentation folder of a dataset (don't forget to add the mapping name in the datasource.json):
  ```json
  {
    "name": "colors",
    "classes": [[1, 2, 3], [4, 5, 6], [7, 8, 9]],
    "colors": [0.2, 0.5, 0.9],
    "hideUnmappedIds": true
  }
  ```
  - Then open a volume tracing of that dataset, brush using the ids 1, 4, and 7
  - Activate the mapping using `mappings.activate("colors")`
  - The respective segments should have the specified colors

- Mappings without custom colors should work as before
- Disabling/Enabling mappings should work as before


### Issues:
- fixes #2173

------
